### PR TITLE
rgw: support object transition in lifecycle config set.

### PR DESCRIPTION
**Pull request by zhangsw**
This feature is useful when rgw supports multiple storage class. 
For more details about S3 transition: https://docs.aws.amazon.com/AmazonS3/latest/dev/intro-lifecycle-rules.html
Signed-off-by: Zhang Shaowen <zhangshaowen@cmss.chinamobile.com>

### CODE REVIEW

File: src/rgw/rgw_lc_s3.h [hunk 1]
```diff
@@ -258,5 +327,5 @@ class RGWLifecycleConfiguration_S3 : public RGWLifecycleConfiguration, public XM
   void dump_xml(Formatter *f) const;
 };
 
-
+bool check_date(const string& _date);
```
Comment (cbodley) [inline src/rgw/rgw_lc_s3.h]: this doesn't need to be exposed in the header if it's only used locally in rgw_lc_s3.cc

File: src/rgw/rgw_lc.h [hunk 2]
```diff
@@ -35,6 +35,13 @@ typedef enum {
   lc_complete,
 }LC_BUCKET_STATUS;
 
+typedef enum {
+  standard_ia = 0,
+  onezone_ia,
+  glacier,
+  undefined,
+} STORAGE_CLASS;
```
Comment (cbodley) [inline src/rgw/rgw_lc.h]: i have a feeling that we'll want to support storage classes with arbitrary names, so i don't think they should be fixed at compile-time with an enum

**Comment by cbodley** [comment]
cc @dang @yehudasa

File: src/rgw/rgw_lc.h (hunk 2 above)
Comment (zhangsw) [reply src/rgw/rgw_lc.h]: em.. I've removed the enum def.

**Comment by cbodley** [comment]
please don't kill this, friendly robot

File: src/rgw/rgw_lc_s3.cc [hunk 3]
```diff
@@ -102,6 +109,51 @@ bool LCFilter_S3::xml_end(const char* el) {
   return !(single_cond && num_conditions > 1);
 }
 
+bool LCTransition_S3::xml_end(const char* el) {
```
Comment (yehudasa) [inline src/rgw/rgw_lc_s3.cc]: We shouldn't use the old style xml decoding api. See use of `RGWXMLDecoder::decode_xml()`, e.g., at `rgw_xml_enc.cc`

File: src/rgw/rgw_lc_s3.cc [hunk 4]
```diff
@@ -269,6 +352,14 @@ XMLObj *RGWLCXMLParser_S3::alloc_obj(const char *el)
     obj = new LCMPExpiration_S3();
   } else if (strcmp(el, "DaysAfterInitiation") == 0) {
     obj = new LCDays_S3();
+  } else if (strcmp(el, "StorageClass") == 0) {
```
Comment (yehudasa) [inline src/rgw/rgw_lc_s3.cc]: I think that while we're at it we should just switch everything to the new style. Would make code more readable and less prone to errors.

File: src/rgw/rgw_lc.h [hunk 5]
```diff
@@ -89,7 +89,68 @@ class LCExpiration
     return true;
   }
 };
-WRITE_CLASS_ENCODER(LCExpiration)
+WRITE_CLASS_ENCODER(LCExpiration);
```
[the remark this answers is not in the document]
Comment (zhangsw) [reply src/rgw/rgw_lc.h]: done

File: src/rgw/rgw_lc.h [hunk 6]
```diff
@@ -89,7 +89,68 @@ class LCExpiration
...
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(days, bl);
+    decode(date, bl);
+    decode(storage_class, bl);
+    DECODE_FINISH(bl);
+  }
+  void dump(Formatter *f) const;
+};
+WRITE_CLASS_ENCODER(LCTransition);
```
[the remark this answers is not in the document]
Comment (zhangsw) [reply src/rgw/rgw_lc.h]: done

File: src/rgw/rgw_lc.h [hunk 7]
```diff
@@ -227,10 +298,20 @@ class LCRule
     dm_expiration = _dm_expiration;
   }
 
+  bool add_transition(LCTransition* _transition) {
```
[the remark this answers is not in the document]
Comment (zhangsw) [reply src/rgw/rgw_lc.h]: done

File: src/rgw/rgw_lc.h [hunk 8]
```diff
@@ -259,13 +342,25 @@ class LCRule
...
      DECODE_FINISH(bl);
    }
   void dump(Formatter *f) const;
 
 };
 WRITE_CLASS_ENCODER(LCRule)
 
+struct transition_action
+{
+  int days;
```
[the remark this answers is not in the document]
Comment (zhangsw) [reply src/rgw/rgw_lc.h]: done

File: src/rgw/rgw_lc_s3.cc (hunk 3 above)
Comment (zhangsw) [reply src/rgw/rgw_lc_s3.cc]: done

File: src/rgw/rgw_lc_s3.cc (hunk 4 above)
Comment (zhangsw) [reply src/rgw/rgw_lc_s3.cc]: Yes， I‘ve switched lifecycle xml decode to the new style.

**Comment by zhangsw** [comment]
@yehudasa How is this pr now？

**Comment by yehudasa** [comment]
@zhangsw oops, we both implemented the same thing, see my branch here:
https://github.com/yehudasa/ceph/commits/wip-rgw-tiering-3

**Pull request closed without merging**